src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -551,6 +553,68 @@ uct_rc_mlx5_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
     return UCS_OK;
 }
 
+static ucs_status_t uct_rc_mlx5_ep_check_cb(uct_pending_req_t *self)
+{
+    uct_rc_pending_req_t *req = ucs_derived_of(self, uct_rc_pending_req_t);
+    uct_rc_mlx5_ep_t *ep      = ucs_derived_of(req->ep, uct_rc_mlx5_ep_t);
+    ucs_status_t status;
+
+    ucs_assert(ep->super.flags & UCT_RC_EP_FLAG_KEEPALIVE_POSTED);
+
+    status = uct_rc_mlx5_ep_check_internal(req->ep);
+    if (status == UCS_OK) {
+        ep->super.flags &= ~UCT_RC_EP_FLAG_KEEPALIVE_POSTED;
+        ucs_mpool_put(req);
+    }
+
+    return status;
+}
+
+ucs_status_t
+uct_rc_mlx5_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
+{
+    UCT_RC_MLX5_EP_DECL(tl_ep, iface, ep);
+    uct_rc_pending_req_t *req;
+    ucs_status_t status;
+
+    UCT_CHECK_PARAM(comp == NULL, "Unsupported completion on ep_check");
+    UCT_CHECK_PARAM(flags == 0, "Unsupported flags: %x", flags);
+
+    if (ep->super.flags & UCT_RC_EP_FLAG_KEEPALIVE_POSTED) {
+        /* keepalive request is in pending queue and will be
+         * processed when resources are available */
+        return UCS_OK;
+    }
+
+    status = uct_rc_mlx5_ep_check_internal(tl_ep);
+    if (status != UCS_ERR_NO_RESOURCE) {
+        ucs_assert(status == UCS_OK);
+        return status;
+    }
+
+    /* there is no iface resources, add pending request */
+
+    /* create mempool for pending requests */
+    status = uct_rc_iface_init_pending_mp(&iface->super);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    req = ucs_mpool_get(&iface->super.tx.pending_mp);
+    if (req == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    req->ep         = &ep->super.super.super;
+    req->super.func = uct_rc_mlx5_ep_check_cb;
+    status          = uct_ep_pending_add(tl_ep, &req->super, 0);
+    if (status == UCS_OK) {
```

Comment:
IMO we should assert_always that status is OK: since uct_rc_mlx5_ep_check_internal returned NO_RESOURCE